### ISSUE TITLE
CI: Reject `ubiquibot-config.yml` changes from unauthorized contributors

### ISSUE THREAD

**Issue by 0x4007** [ubiquity/ubiquibot#836]
Only `admins` and `billing_managers` should have the authority to change the bot config due to price sensitivity. Specifically anything related to finances. 

I quickly parsed out the finance related properties from the default config:

```typescript
{
  evmNetworkId: 100,
  priceMultiplier: 1,
  issueCreatorMultiplier: 2,
  paymentPermitMaxPrice: 9007199254740991,
  assistivePricing: false,
  commentIncentives: false,
  incentives: {
    comment: {
      elements: {},
      totals: {
        word: 0,
      },
    },
  },
  enableAccessControl: {
    label: false,
    organization: true,
  },
}
```

**Comment by BeanieMen** [comment]
How would it reject it?

**Comment by Keyrxng** [comment]
I'll draft this ~~tomorrow~~ today

**Comment by 0x4007** [comment]
We did pause merges on this repo until the refactor is merged in though. But at least it'll allow you to get a head start on the research. You might have merge conflicts though.

**Comment by Keyrxng** [comment]
> We did pause merges on this repo until the refactor is merged in though.

I noticed that but I need to work so try and stop me :))

> You might have merge conflicts though.

On this one I think I'll get away with but the others I have dormant, definitely will have a host of them but I appreciate the heads-up

**Comment by Keyrxng** [comment]
@pavlovcik or @rndquu, the roles come from the org right? Not the db?

**Comment by rndquu** [comment]
> @pavlovcik or @rndquu, the roles come from the org right? Not the db?

yes

**Comment by Keyrxng** [comment]
> yes

figured and a few mins after I asked found where to inv myself 😂 cheers mate

**Comment by Keyrxng** [comment]
Tougher than I expected but nearly there

I'm using a dummy account to QA and it seems that ``billing_manager`` is overridden by any other role as opposed to listing all applicable as the ``role`` prop is an enum. So I'm trying to think of edge cases but I've little behind the scenes info. What I mean by that is, are ``billing_manager``'s active contributors or are they more 'back office'? It'll need covered but just trying to think of workarounds for the below.

What I've tested so far, for instance:

- In a private repo the billing manager would need invited and therefor would have a role assigned as part of the invite for that repo in particular

- At an org level, they cannot be a member otherwise it overrides but this in turn stops the api needed to actually fetch the role info.

I'll get there but if you've any insight that would help I'd appreciate it

**Comment by Keyrxng** [comment]
Q & A:

- Rest API: As far as I can see, the only way to return billing_manager role is through the teams api, so are they assigned a team typically?

- Rest API: It seems like you can pull it via the invitation endpoint but pending invites expire and accepted ones aren't kept so that's a bust, not a really a question either

- GraphQL [Doc link](https://docs.github.com/en/graphql/reference/objects#orgaddbillingmanagerauditentry): "Organization billing plan does not support access to auditLog information" So we can't use auditLog, unless I'm mistaken and UBQ would have access to that?

- Are billing_managers often an admin as well, I'd imagine so but reading the github docs throw me off as they abstract the billing manager away from being directly involved.

<hr/>

Basically then it seems like checking for billing_manager status is out of the question unless it's a paid Enterprise, unless they are added to the org as an admin as well I think I'll only be able to check for admin status on config updates.

Please correct me if I'm wrong

**Comment by Keyrxng** [comment]
The config, I assume for a particular reason, doesn't use double quotes which is making it hard to parse in order to do a granular review of changed properties.

- Is this a necessity that it be so granular? 
- Or is it sufficient enough to just do: "isAdmin ? pass : fail"

The point being here, whenever someone without authority changes the config we want to be warned the config has changed. However, in cases where the change is required the CI will need to be confirmed by the reviewers ANYWAY, so in all scenarios it fails and would continue to fail unless committed by an admin or overridden as an acceptable change. Considering this, "isAdmin ? pass : fail" makes sense to me.

**Comment by Keyrxng** [comment]
Forgot about the other action that updates the config. Got it working against my private org repo, draft it maybe tonight probably tomorrow

**Comment by 0x4007** [comment]
I've got mixed feelings on the yml implementation. Installation to partner repos will be more complicated if they need to add yml. 

Instead ideally the bot should handle it in its core code so that when they add the bot to their repo, it can handle this use case. 

I haven't come to a conclusion on this because I asked @wannacfuture to research delegating long running processes to GitHub actions (permit generation and AI related tasks) which might require us to figure some flow that allows yml installation when the bot is added to the repo etc.

**Comment by Keyrxng** [comment]
At first I did start building it into the Bot yknow and spoke myself out of it with incorrect thinking that it would be easier as an action but it would have been simpler probably, that's why I ask the below, which itself doesn't makes sense but my train of thought was all wrong at first lmao

> pavlovcik or rndquu, the roles come from the org right? Not the db?

So for now just sit on things until the research comes in, not a problem

**Comment by 0x4007** [comment]
> So for now just sit on things until the research comes in, not a problem

@wannacfuture any remarks to add?

**Comment by 0x4007** [comment]
Remind me after the refactor and we'll get reviews going again etc.

**Comment by BeanieMen** [comment]
Yay
